### DISCUSSION TITLE
Unify outputs

### DISCUSSION THREAD

**Discussion by augustoteixeira** [cartesi/cips#21] [General]
This CIP was a suggestion by @diegonehab and it unifies Vouchers and Notices on the low level implementation (read Server Manager):

https://github.com/cartesi/cips/blob/unify-output/cips/cip-unify-output.md

One of the main advantages of this is the ability to define new types of outputs (besides Vouchers and Notices) without requiring any changes on the Server Manager.

**Comment by augustoteixeira** [comment]
Hi @diegonehab and @vfusco, do you think this will have a positive impact on the Server Manager architecture?

**Reply by diegonehab** [reply]
Certainly positive impact (by positive I mean reduction in complexity and lines of code). The gRPC interface will become simpler. There will be a some reduction in code size. The biggest change is in the Cartesi Machine. There we will remove one of the memory ranges (notice-hashes and voucher-hashes -> hashes) everywhere it appears in the code and API.

**Comment by felipeargento** [comment]
I like this a lot! 

It simplifies the onchain code, simplifies the proof structure and also makes it easier to reason about the possible repercussions of a Cartesi in L1. Great initiative.

I think we should try to prioritize this and try to include this CIP on the next release, provided that it doesn't create too much hassle on the Server Manager architecture.

One question: can't we, instead of adding 28 bytes of padding, concatenate the length and header? I don't think we need 32 bytes for the length. All gas saved on decoding (even if it is a single extra word in calldata) is room to expand the flexibility of vouchers.

We also have to be sure that the voucher encoding is compliant with the solidity abi, to avoid onchain processing: https://docs.soliditylang.org/en/v0.8.13/abi-spec.html?highlight=encode#function-selector-and-argument-encoding

I can already imagine many different types of useful outputs to propose (in new CIPs)!

**Reply by augustoteixeira** [reply]
> I think we should try to prioritize this and try to include this CIP on the next release, provided that it doesn't create too much hassle on the Server Manager architecture.

Let us try to sync with @diegonehab and @vfusco when it is the best release to include this. But I agree that it deserves to have a priority. The automatic execution of outputs would greatly benefit from this.

> One question: can't we, instead of adding 28 bytes of padding, concatenate the length and header? I don't think we need 32 bytes for the length. All gas saved on decoding (even if it is a single extra word in calldata) is room to expand the flexibility of vouchers.

This is something I keep thinking about. For the inputs, we should compress as much as possible (because decoding is free and data availability is expansive). But for outputs both costs enter the picture.

I was basing my suggestion in the first warning of this documentation:
https://docs.soliditylang.org/en/v0.8.13/internals/layout_in_storage.html

But it is better to leave this decision for later once it is being implemented. Perhaps it is easy to measure the cost of the alternatives.

> I can already imagine many different types of useful outputs to propose (in new CIPs)!

:)

**Reply by diegonehab** [reply]
I am agnostic to the format, though I would like us to use the same "frame" format for inputs and outputs. I'd like to have a length field first, and whatever else you guys want could come next. We could even do something "creative" and have that length field not be part of the hash, so you guys don't receive it on-chain. This is doable. I don't care if the length is 4 bytes only.

**Reply by diegonehab** [reply]
The reason I'd like to have a frame for inputs and outputs is so we can use a "layered" approach to decoding. The lowest levels would only understand size-prefixed-blobs. On top of it, we would write user-space code that would interpret what's inside. Then on top of that, we could add more layers etc.

**Reply by augustoteixeira** [reply]
The problem with this is that decoding vouchers has to be done by a smart contract on L1 (this is expensive, specially if not following Solidity's standards). So perhaps we should try to have as much as a "layered" approach as we possibly can, but without making it expensive to decode on-chain.

**Reply by diegonehab** [reply]
Mixing abstraction levels creates a lot of confusion. So let's think this through. What I am suggesting does not seem to affect decoding at all in the blockchain. Just adding a length field right before what is now known as the tx_buffer and what is now know as rx_buffer. Something like that.

**Reply by augustoteixeira** [reply]
Let us discuss this with care indeed, there are several competing things going on here:
- The Server Manager has to be able to retrieve the contents of the output (trimmed at the right length);
- The "splicing" phase of the verification game needs to be able to prove that the contents of the tx_buffer match the "output_hash". **For this it is very important that the size of the `tx_buffer` be limited!!!!**;
- The Output Contract needs to decode the contents of the output corresponding to that hash.
Let us discuss.

**Comment by augustoteixeira** [comment]
This unification does involve work from several fronts:
- the encoding still needs to be decided upon;
- some smart contracts have to be changed (nothing huge);
- the grpc interface changes;
- the Server Manager and HTTP API have to adapt;
- the indexer has to change to adapt to the new grpc;
- the example DApps will have to change slightly as well.

All the changes seem to be reasonably small. But the coordination process could prove difficult. I don't know how to get the ball rolling...

**Reply by vfusco** [reply]
@augustoteixeira I would suggest we start with:

1. Encoding
2. GRPC Interfaces changes
3. OpenAPI Interfaces changes

Then we can create issues to update the rest.

**Reply by augustoteixeira** [reply]
Good starting point!

I will go one step further and detail this a bit more, trying to split the work in three phases:

**Phase 0**
- We discuss with everyone the **Output API** that we are going to use. Look at the discussion above to understand some of the challenges.

**Phase 1**
- In this first phase, we will keep both "vouchers" and "notices" in some parts of the system. Namely: the Server Manager, the Linux Driver and the GraphQL Server. This minimizes the amount of chages in the first phase;
- We change the HTTP API, so that (although it still receives `/vouchers` and `/notices`) it puts both in the driver as if they were Notices (encoded by the **Output API**);
- We change the Output Contract, so that it only reads Notices from now on and it executes notices that are formatted as Vouchers according to the **Output API**;
- We change the Indexer Server, so that it understands the Output API. Then (although there are only notices being produced) it is able to tell what are vouchers or notices and fills the Database correctly.
This should give us a working product. Observe that no change is needed in: the example dapps or their frontends.

**Phase 2**
- We change the Linux Driver, so that now it only accepts generic outputs.
- We change the Server Manager, so that it only handles generic outputs.

What do you think, @felipeargento, @diegonehab, @vfusco?

**Comment by miltonjonat** [comment]
Do you guys see any impacts on the developer-facing APIs (front-end and back-end)? The back-end would still call `/voucher` and `/notice`, for instance?

**Reply by felipeargento** [reply]
I'd say there is very little impact if any.

While they are all `outputs`, it still makes sense for the developer-facing APIs to treat them differently. Vouchers will have an `execution status` that does not make sense for notices, for instance. But this opens the door for us having even more output types or even different types of `vouchers`. I still see developers calling `/voucher` or `/output/voucher` though.

**Reply by diegonehab** [reply]
I think we should keep providing the direct entrypoints for `voucher` and `notice`, but we could also provide a more general entrypoint that allows for custom `output`.
